Add vtkCleaverImageToUnstructuredGridFilter

aronhelser opened the pull request:
This initial version does something useful. It provides the same parameters and defaults as the ITK filter that also wraps Cleaver, and constructs an unstructured grid output directly from the Cleaver result mesh. 

The test creates a very course image, with two spherical regions with values 1.0 and 2.0 - a labelmap. To transform the labelmap into the indicator functions Cleaver wants, I mimicked an ITK distance filter by combining two VTK distance filters for inside and outside the region. The test checks for the expected number of cells and points in the output, and has a rendered image baseline comparison. Here's the output:
![vtkCleaverImageToUnstructuredGridFilterTest](https://user-images.githubusercontent.com/19917213/162092419-843116df-74db-41d6-af86-3c6dc950ab58.png)

The padding parameter and indicator function toggle still need to be hooked up.

aronhelser commented:
Next, working on creating unstructured grid output directly instead of writing a file. Then wiring up parameters.

aronhelser commented:
@jcfr this filter is minimally functional now. Still need to hook up padding and indicator function flag. We can start working on integration with something else now...

jcfr commented on vtkCleaverImageToUnstructuredGridFilter.h:
```diff
@@ -0,0 +1,107 @@
+#ifndef vtkCleaverImageToUnstructuredGridFilter_h
+#define vtkCleaverImageToUnstructuredGridFilter_h
+
+#if defined(_WIN32)
+#define VTKCLEAVER_EXPORT __declspec( dllexport )
+#else
+#define VTKCLEAVER_EXPORT
+#endif
```
Remove this and instead include 

```
#include "vtkCleaverImageToUnstructuredGridModule.h" // For export macro
```

aronhelser replied:
OK, I thought we had to include cmake code for this to be available, but I see the vtk module takes care of it.

jcfr replied:
> but I see the vtk module takes care of it.

Indeed, with the current approach everything is taking care of and the next step will only be one possible approach for dissemination.

aronhelser replied:
fixed.

jcfr commented on vtk.module:
```diff
@@ -0,0 +1,32 @@
+NAME
+  VTK::CleaverImageToUnstructuredGrid
```
```suggestion
  VTK::Cleaver
```

aronhelser replied:
done

jcfr commented on vtk.module:
```diff
@@ -0,0 +1,32 @@
+NAME
+  VTK::CleaverImageToUnstructuredGrid
+LIBRARY_NAME
+  vtkCleaverImageToUnstructuredGrid
```
```suggestion
  vtkCleaver
```

jcfr commented on vtkCleaverImageToUnstructuredGridFilter.cxx:
```diff
@@ -0,0 +1,409 @@
+    // distance map
+    distance[jj]->SetInputConnection(thresholdDistance[jj]->GetOutputPort());
+  }
+  // inverse of each other
```
explain why

aronhelser replied:
done

jcfr commented on vtkCleaverImageToUnstructuredGridFilter.cxx:
```diff
@@ -0,0 +1,409 @@
+    imgdiff->Update();
+    auto* img = imgdiff->GetOutput();
+
+    int dims[3];
```
I suggest to use `vtkIdType`

jcfr reviewed (commented):
Few minor nitpicks.

Could you also include screenshot of input and output in the pull request description ?

aronhelser commented:
I'm used to gitlab - there's not a pull request description at the top, or am I missing something?

The test input is generated in the test, and it's a volume, so no easy screen shot - 24 x 24 x 24 with two partial spherical labels with values 1.0 and 2.0. 

Here's the rendered output:
![test](https://user-images.githubusercontent.com/19917213/162063643-daf3b281-4678-42b4-b427-45a130032cc0.png)

jcfr commented:
> there's not a pull request description at the top, or am I missing something?

![image](https://user-images.githubusercontent.com/219043/162069590-c94fdabe-caed-4614-a252-3f1b8aa1a5de.png)

By editing the description you can then provide a complete summary of all changes.

aronhelser commented:
OK, using the initial comment makes sense. gitlab has a separate description from the comment stream.

aronhelser commented on vtkCleaverImageToUnstructuredGridFilter.h:
```diff
@@ -0,0 +1,107 @@
+  ///@}
+
+protected:
+    vtkCleaverImageToUnstructuredGridFilter();
```
reformatted with clang-format

aronhelser commented on vtkCleaverImageToUnstructuredGridFilter.h:
```diff
@@ -0,0 +1,107 @@
+
+protected:
+    vtkCleaverImageToUnstructuredGridFilter();
+    ~vtkCleaverImageToUnstructuredGridFilter();
```
done

aronhelser commented on vtkCleaverImageToUnstructuredGridFilter.cxx:
```diff
@@ -0,0 +1,409 @@
+  };
+typedef std::map< const vec3, size_t, vec3_compare > VertMap;
+// given a cleaver TetMesh, fill in the UnstructuredGrid output. Based on
+// TetMesh::writeVtkUnstructuredGrid()
```
done

aronhelser commented on vtkCleaverImageToUnstructuredGridFilter.cxx:
```diff
@@ -0,0 +1,409 @@
+      vtkLog(INFO, "Md " << md );
+
+      // ITK itkApproximateSignedDistanceMapImageFilter has inside/outside values
+      // for VTK, anything non-zero is considered NOT an object. Threshold again to
```
I added more comments about why its two parallel pipelines, one for calculating distance outside the label, and one for inside. Then they are combined with subtraction.

aronhelser commented on vtkCleaverImageToUnstructuredGridFilter.cxx:
```diff
@@ -0,0 +1,409 @@
+  vtkNew<vtkImageEuclideanDistance> distance[2];
+  vtkNew<vtkImageThreshold> thresholdDistance[2];
+  vtkNew<vtkImageMathematics> imgdiff;
+  for(auto jj = 0; jj < 2; ++jj)
```
changed to idx

aronhelser commented:
Thanks for the detailed review - keeping me honest when borrowing code from different tests.
I think I've addressed all comments, but I could have missed something. Do you want to take another look @jcfr ?

aronhelser commented:
Added support for indicator function inputs. Padding support was already added. Features are now at parity with the ITK filter. Here's the second test image:
![image](https://user-images.githubusercontent.com/19917213/162796456-11ed3ac6-28db-4933-aeea-c64dd9dc849e.png)
